engine/runtime/src/main/java/org/enso/interpreter/runtime/Module.java

```diff
@@ -154,15 +160,18 @@ public Module(QualifiedName name, Package<TruffleFile> pkg, Rope literalSource)
    * @param pkg the package this module belongs to. May be {@code null}, if the module does not
    *     belong to a package.
    */
-  private Module(QualifiedName name, Package<TruffleFile> pkg) {
-    this.sources = ModuleSources.NONE;
+  private Module(
+      QualifiedName name, Package<TruffleFile> pkg, boolean virtual, Optional<Rope> literalSource) {
+    this.sources =
+        literalSource.map(s -> ModuleSources.NONE.newWith(s)).orElseGet(() -> ModuleSources.NONE);
     this.name = name;
     this.scope = new ModuleScope(this);
     this.pkg = pkg;
-    this.compilationStage = CompilationStage.AFTER_CODEGEN;
+    this.compilationStage = virtual ? CompilationStage.INITIAL : CompilationStage.AFTER_CODEGEN;
```

Comment:
Virtual modules, that's something @4e6 may also be interested in for his _lazy visualizations_.